ProdLDA example with Haiku + Flax

tcbegley opened the pull request:
First version of NumPyro port of [ProdLDA notebook from Pyro docs](http://pyro.ai/examples/prodlda.html) as requested in #1033. It's branched off #1016.

Having some issues with unexpected tracers inside SVI when differentiating the loss function. I think it's to do with the batchnorm mutable state, but it's a bit beyond my understanding atm. Any pointers would be appreciated.

tcbegley commented on notebooks/source/prodlda.ipynb:
```diff
@@ -0,0 +1,794 @@
+    "rng_key = random.PRNGKey(0)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "nbsphinx": "hidden",
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "import os\n",
+    "\n",
+    "assert numpyro.__version__.startswith(\"0.6.0\")\n",
+    "# Enable smoke test - run the notebook cells on CI.\n",
+    "smoke_test = \"CI\" in os.environ"
+   ]
+  },
```
Adding

```
{
    "nbsphinx": "hidden",
    "tags": []
}
```

to the cell metadata allows you to suppress cells from the rendered output. I think this could be a nice pattern for removing smoke test configuration from the docs, but still allowing notebooks to be run in CI. What do you think?

fehiepsi reviewed (commented):
Hi @tcbegley, I just fixed a bug in #1016 that caused the issue. Sorry for that!

I think it is better to make an example for prodlda that links to the original tutorial for a few reasons:
+ only code is changed I guess - the content seems to be similar
+ the main attribute should be given to the Pyro tutorial
+ we can add [comments like this](https://github.com/pyro-ppl/numpyro/blob/master/examples/capture_recapture.py#L57); sphinx-gallery will render the text well if some explanation for haiku/flax is needed. You can also add the result image, like in [this example](https://github.com/pyro-ppl/numpyro/blob/4ad6b98ae3127e0d509ca7ff425cbf07185fbdad/examples/funnel.py#L25)
+ for comparison - it might be better to look at an end-to-end code in numpyro, without the need to follow a similar training pattern in Pyro
+ it is easier to add flax code :D

 About the implementation, most look great to me. I just have a few comments about the API:
+ making `HaikuEncoder` seems to be unnecessary - I guess you can just default
```python
def HaikuEncoder(x, is_training, num_topics, hidden, dropout_rate):
    # follow the same implementation of HaikuEncoder.__call__

def model(docs, hypers, is_training=False):
    # hyperparameters can be accessed through hypers["vocal_size"],...
    encoder = haiku_module(
        "encoder",
        hk.transform_with_state(partial(HaikuEncoder, num_topics=...)),
        ...
```

The data is not large so you can use [svi.run](https://num.pyro.ai/en/latest/svi.html#numpyro.infer.svi.SVI.run) to train, just make sure that you use `numpyro.subsample` correctly
```python
def model(docs, ...):
    with numpyro.plate("documents", docs.shape[0], subsample_size=hyper["batch_size"]):
        logtheta = numpyro.sample(
            "logtheta", dist.Normal(0, 1).expand([num_topics]).to_event(1)
        )
        theta = jax.nn.softmax(logtheta)
        ...
        batch_docs = numpyro.subsample(docs, event_dim=1)
        numpyro.sample("obs", ..., obs=batch_docs)
```

tcbegley commented:
Thanks for the review!

Been a bit busy lately so progress is a little slow, but have addressed your comments:
- Drop notebook in favour of script
- Not subclassing `hk.Module` in Encoder / Decoder implementations
- Ditching extra functional wrappers such as `build_model_and_guide`.
- Use `svi.run` with `numpyro.subsample`.

The script runs without error, but in my testing the model doesn't actually learn anything useful. Will need to do some experimentation to figure out why exactly. If you see anything obviously wrong let me know!

TODO: 
- [ ] Resolve training issues
- [ ] Allow variables to be overridden from the commandline (i.e. use `argparse` for setting hyperparameters)
- [ ] Add Flax
- [ ] Investigate using Dirichlet distribution instead of softmax normal as suggested in #1033 (I figured I'd try and match outputs of the Pyro implementation before trying this)

fehiepsi commented:
> Resolve training issues

Thanks, @tcbegley! That's a great plan and the script is pretty clean!! I'll look at this issue today.

fehiepsi commented on examples/prodlda.py:
```diff
@@ -0,0 +1,207 @@
+    def __call__(self, inputs, is_training):
+        dropout_rate = self._dropout_rate if is_training else 0.0
+        h = hk.dropout(hk.next_rng_key(), dropout_rate, inputs)
+        h = hk.Linear(self._vocab_size)(h)
```
I think we need `with_bias=False` here.

fehiepsi commented on examples/prodlda.py:
```diff
@@ -0,0 +1,207 @@
+        dropout_rate = self._dropout_rate if is_training else 0.0
+        h = hk.dropout(hk.next_rng_key(), dropout_rate, inputs)
+        h = hk.Linear(self._vocab_size)(h)
+        h = hk.BatchNorm(create_scale=True, create_offset=True, decay_rate=0.1)(
```
You can set `create_scale=False` and `create_offset=False`

tcbegley replied:
Thanks! I suspected I had got something wrong with my batchnorms >_<

fehiepsi commented on examples/prodlda.py:
```diff
@@ -0,0 +1,207 @@
+        logtheta_loc = jnp.zeros((batch_docs.shape[0], hyperparams["num_topics"]))
+        logtheta_scale = jnp.ones((batch_docs.shape[0], hyperparams["num_topics"]))
+        logtheta = numpyro.sample(
+            "logtheta", dist.Normal(logtheta_loc, logtheta_scale).to_event(1)
```
You can use `dist.Normal(0, 1).expand([hyperparams["num_topics"]]).to_event(1)` for simplicity.

tcbegley replied:
Very slick 🙂

fehiepsi commented on examples/prodlda.py:
```diff
@@ -0,0 +1,207 @@
+
+        count_param = decoder(numpyro.prng_key(), theta, is_training)
+
+        total_count = batch_docs.sum(-1).max().astype(int)
```
This should be `total_count = batch_docs.sum(-1)`

fehiepsi commented on examples/prodlda.py:
```diff
@@ -0,0 +1,207 @@
+    docs = device_put(docs)
+    batch_size = 32
+    learning_rate = 1e-3
+    num_steps = 200_000
```
I think 30,000 is enough (similar to the number in Pyro).

tcbegley replied:
I think I bumped this up to see what would happen when it wasn't learning anything in my initial testing and forgot I had done that 😳

fehiepsi reviewed (commented):
@tcbegley Could you double-check if the following fixes solve the issue? I am seeing similar losses w.r.t. Pyro tutorial (note that in Pyro, the reported running loss is scaled down by `batch_size`).

fehiepsi commented on examples/prodlda.py:
```diff
@@ -0,0 +1,207 @@
+    optimizer = numpyro.optim.Adam(learning_rate)
+    svi = SVI(model, guide, optimizer, loss=TraceMeanField_ELBO())
+
+    params, losses = svi.run(rng_key, num_steps, docs, hyperparams, is_training=True)
```
new api: `svi_result = svi.run(...)`

fehiepsi commented on examples/prodlda.py:
```diff
@@ -0,0 +1,207 @@
+
+    params, losses = svi.run(rng_key, num_steps, docs, hyperparams, is_training=True)
+
+    plt.plot(losses)
```
this is `svi_result.losses`

fehiepsi commented on examples/prodlda.py:
```diff
@@ -0,0 +1,207 @@
+
+    plt.plot(losses)
+
+    beta = params["decoder$params"]["haiku_decoder/linear"]["w"]
```
this is `beta = svi_result.params["decoder$params"]["linear"]["w"]

fehiepsi commented on examples/prodlda.py:
```diff
@@ -0,0 +1,207 @@
+        )
+
+
+def plot_word_cloud(b, ax, n):
```
the arguments are `plot_word_cloud(b, ax, vocab, n)`

tcbegley replied:
Oops, good catch!

fehiepsi commented:
I can confirm that with the above change, we get the desired result.

fehiepsi commented on examples/prodlda.py:
```diff
@@ -0,0 +1,332 @@
+much more quickly. Furthermore, it does not require a custom inference algorithm that
+relies on complex mathematical derivations. This example also serves as an
+introduction to Flax and Haiku modules in NumPyro.
+
```
Could you add a paragraph "For the interested readers, a nice extension of this model is the CombinedTM model (reference [3]) which utilizes a pre-trained sentence transformer (like https://www.sbert.net/) to generate a better representation of the encoded latent vector." ?

fehiepsi reviewed (commented):
Awesome work, @tcbegley! The implementation is so clean (I can quickly learn the difference between flax and haiku :D). This is a great example even for users who are not familiar with those nn frameworks. +1

tcbegley commented on examples/prodlda.py:
```diff
@@ -0,0 +1,332 @@
+        return nn.softmax(h)
+
+
+def model(docs, hyperparams, is_training=False, use_haiku=False):
```
I went with `nn_framework` rather than `nn_module` because I thought the latter might get confused with `nn.Module`? Probably doesn't matter but let me know if you would prefer `nn_module`.

fehiepsi replied:
Yeah, agree. I like `nn_framework`. :D

tcbegley commented on examples/prodlda.py:
```diff
@@ -0,0 +1,332 @@
+
+def plot_word_cloud(b, ax, vocab, n):
+    indices = jnp.argsort(b)[::-1]
+    df = pd.DataFrame(indices[:100], columns=["index"])
```
I tried 20 and it looked a little sparse so I went with 30. What do you think? (See results [here](https://github.com/pyro-ppl/numpyro/blob/45716595bb856879a51d181527e51bebc1dade08/docs/source/_static/img/examples/prodlda.png))

fehiepsi replied:
That looks great, thanks for trying out! (I just thought that 100 is a bit noisy)

tcbegley commented on examples/prodlda.py:
```diff
@@ -0,0 +1,332 @@
+        description="Probabilistic topic modelling with Flax and Haiku"
+    )
+    parser.add_argument("-n", "--num-steps", nargs="?", default=30_000, type=int)
+    parser.add_argument("-t", "--num-topics", nargs="?", default=20, type=int)
```
Went with 12 as the default but adjusted the plotting code so it can handle any number 💪

tcbegley commented on examples/prodlda.py:
```diff
@@ -0,0 +1,360 @@
+        # the number of learning parameters
+        h_mu = hk.Linear(self._num_topics)(h)
+        logtheta_loc = hk.BatchNorm(
+            create_scale=False, create_offset=False, decay_rate=0.99
```
I think this is the same as the default in flax (though their parameter is called `momentum` rather than `decay_rate`).

Haiku requires that you set this, should I set it in the flax layers too to be more explicit? Also is `0.99` a good choice for this problem? I don't have that much intuition for batch norm...

fehiepsi replied:
Good point! Luckily we didn't perform any prediction in this example, so the running mean/scale is not used anywhere. 

> intuition for batch norm

Currently, `momentum=0.1` is used by default in PyTorch [BatchNorm1d](https://pytorch.org/docs/stable/generated/torch.nn.BatchNorm1d.html#torch.nn.BatchNorm1d) (also in [ctm model](https://github.com/MilaNLProc/contextualized-topic-models/blob/master/contextualized_topic_models/networks/decoding_network.py#L82)) to put less weights on the new input value when updating. The `momentum` notion in PyTorch is equivalent to `1 - momentum` notion in Flax/Haiku, so I think it is fine to use `0.9` here. I guess `0.99` also works but changing the default value in flax will make that hyperparameter more explicit to reader :D

tcbegley commented:
Thanks for reviewing again @fehiepsi!

I think I've addressed all your comments. Still have the Dirichlet distribution to try out, but I think then I'll be more or less done.

fehiepsi reviewed (commented):
Yeah, I think we can use Dirichlet distribution directly in future iterations. Could you add this example to the applications section of [this index file](https://github.com/pyro-ppl/numpyro/blob/master/docs/source/index.rst)? Could you also add tests for both `flax` and `haiku` [here](https://github.com/pyro-ppl/numpyro/blob/master/test/test_examples.py#L39) (with small `num_steps` and `hidden`)? Then I think this is ready. Fantastic work, @tcbegley!!!

tcbegley commented:
> I think we can use Dirichlet distribution directly in future iterations.

To clarify, do you mean I should not try that on this PR and leave that for a future update?

fehiepsi commented on setup.py:
```diff
@@ -65,7 +65,14 @@
         ],
-        "examples": ["arviz", "jupyter", "matplotlib", "pandas", "seaborn"],
+        "examples": [
+            "arviz",
```
I think you'll also need `scikit-learn` to get data.

fehiepsi commented on test/test_examples.py:
```diff
@@ -37,6 +37,7 @@
     "minipyro.py",
     "neutra.py --num-samples 100 --num-warmup 100",
     "ode.py --num-samples 100 --num-warmup 100 --num-chains 1",
+    "prodlda.py --num-steps 10 --hidden 10",
```
Could you test for both flax and haiku?
```
"prodlda.py --num-steps 10 --hidden 10 --nn_framework='flax'",
"prodlda.py --num-steps 10 --hidden 10 --nn_framework='haiku'",
```

fehiepsi commented:
> should not try that on this PR and leave that for a future update

Sorry, I thought you wanted to defer it to a future PR. It would be nice to try in this PR. To make the training process more stable, could you register KL divergence for [Dirichlet-Dirichlet](https://github.com/pytorch/pytorch/blob/05c1e5b65571d2fa14b0f06c37ba970fd7cc43d0/torch/distributions/kl.py#L232) pair [here](https://github.com/pyro-ppl/numpyro/blob/f8f482ad7b2f3f658d198c99f76fdda12cff84ca/numpyro/distributions/kl.py#L198)?

tcbegley commented:
> It would be nice to try in this PR.

Oh great, I misunderstood. I'll take a more careful look into it then 🙂

fehiepsi commented on examples/prodlda.py:
```diff
@@ -0,0 +1,369 @@
+            momentum=0.9,
+            use_running_average=not is_training,
+        )(h)
+        return nn.softmax(h)
```
It is more stable to return `h` here and use `Multinomial(..., logits=...)`.

fehiepsi commented on examples/prodlda.py:
```diff
@@ -0,0 +1,369 @@
+        if nn_framework == "flax":
+            count_param = decoder(
+                theta, is_training, rngs={"dropout": numpyro.prng_key()}
+            )
```
It is more stable to define `logits = decoder(...)`

fehiepsi commented on examples/prodlda.py:
```diff
@@ -0,0 +1,369 @@
+
+        total_count = batch_docs.sum(-1)
+        numpyro.sample(
+            "obs", dist.Multinomial(total_count, count_param), obs=batch_docs
```
use `Multinomial(total_count, logits=logits)`

fehiepsi commented on examples/prodlda.py:
```diff
@@ -0,0 +1,369 @@
+            "logtheta",
+            dist.Normal(0, 1).expand((hyperparams["num_topics"],)).to_event(1),
+        )
+        theta = jax.nn.softmax(logtheta)
```
This can be changed into
```diff
-        logtheta = numpyro.sample(
-            "logtheta",
-            dist.Normal(0, 1).expand((hyperparams["num_topics"],)).to_event(1),
+        theta = numpyro.sample(
+            "theta",
+            dist.Dirichlet(jnp.ones(hyperparams["num_topics"]))
,
         )
-        theta = jax.nn.softmax(logtheta)
```

fehiepsi commented on examples/prodlda.py:
```diff
@@ -0,0 +1,369 @@
+            )
+
+        numpyro.sample(
+            "logtheta", dist.Normal(logtheta_loc, logtheta_scale).to_event(1)
```
My dirichlet version
```diff
         if nn_framework == "flax":
-            logtheta_loc, logtheta_scale = encoder(
+            concentration = encoder(
                 batch_docs, is_training, rngs={"dropout": numpyro.prng_key()}
         ...
-            "logtheta", dist.Normal(logtheta_loc, logtheta_scale).to_event(1)
+            "theta", dist.Dirichlet(concentration)
```

fehiepsi reviewed (commented):
Here are my changes to use Dirichlet distribution directly. It seems to work but you might try a higher `num_steps` or other hyperparameters.

tcbegley commented on examples/prodlda.py:
```diff
@@ -0,0 +1,369 @@
+            use_running_average=not is_training,
+        )(h_sigma)
+        logtheta_scale = jnp.exp(0.5 * logtheta_logvar)
+        return logtheta_loc, logtheta_scale
```
Thanks! I think when I tried briefly the other day I forgot to exponentiate the log concentration 🤦‍♂️

tcbegley commented:
Thanks for the pointers @fehiepsi! Registering the KL like that is really neat.

Tested a few times with different hyperparameters. I don't think the `num_steps` needs increasing, if anything I think you could get away with decreasing it! I did increase `hidden` as the topics seemed a bit better defined when doing so. When using GPU for inference this had no noticeable impact on runtime.

I'm not sure if the examples tests are going to pass, I ran them locally in a fresh virtual environment, running only the commands that get run in CI and it was fine, I also [ran the prodlda tests on my fork](https://github.com/tcbegley/numpyro/pull/1/files), with the other examples tests disabled so I could read the logs more easily, and in both cases they passed, but they've failed here a couple times in a row. If they fail again I'll take a more careful look into why.

fehiepsi commented on examples/prodlda.py:
```diff
@@ -0,0 +1,346 @@
+        h = nn.softplus(nn.Dense(self.hidden)(inputs))
+        h = nn.softplus(nn.Dense(self.hidden)(h))
+        h = nn.Dropout(self.dropout_rate, deterministic=not is_training)(h)
+
```
nit: no empty space

fehiepsi commented on examples/prodlda.py:
```diff
@@ -0,0 +1,346 @@
+
+        # NB: here we set `create_scale=False` and `create_offset=False` to reduce
+        # the number of learning parameters
+        h = hk.Linear(self._num_topics)(h)
```
nit: move this to the line 61.

fehiepsi commented on examples/prodlda.py:
```diff
@@ -0,0 +1,346 @@
+much more quickly. Furthermore, it does not require a custom inference algorithm that
+relies on complex mathematical derivations. This example also serves as an
+introduction to Flax and Haiku modules in NumPyro.
+
```
Could you mention the change for using Dirichlet directly, instead of the Softmax-Normal approximation here?

fehiepsi reviewed (approved):
LGTM! Could you help me add a test for KL [here](https://github.com/pyro-ppl/numpyro/blob/f8f482ad7b2f3f658d198c99f76fdda12cff84ca/test/test_distributions.py#L2035)? The code should be similar to NormalNormal pair but you might need a higher number of samples than `10000`.

fehiepsi commented on test/test_distributions.py:
```diff
@@ -2033,3 +2033,13 @@ def test_kl_normal_normal(shape):
+@pytest.mark.parametrize("shape", [(), (4,), (2, 3)], ids=str)
+def test_kl_dirichlet_dirichlet(shape):
+    p = dist.Normal(np.exp(np.random.normal(size=shape)))
+    q = dist.Normal(np.exp(np.random.normal(size=shape)))
```
Normal -> Dirichlet

fehiepsi commented on test/test_distributions.py:
```diff
@@ -2033,3 +2033,13 @@ def test_kl_normal_normal(shape):
     assert_allclose(actual, expected, rtol=0.05)
+
+
+@pytest.mark.parametrize("shape", [(), (4,), (2, 3)], ids=str)
```
For dirichlet, you should remove `()` because Dirichlet is a multivariate distribution.
